Adding Support for Complex Watermark Types

sahilTakiar opened the pull request:
This is the first pull request for adding complex watermark types to Gobblin. This will replace the legacy system of tracking watermarks. The old system was de-centralized, and depended and passing custom configuration parameters between executions via a `WorkUnitState`.

The new implementation contains a new interface called `Watermark` which extends the `Comparable` and the `Copyable` interfaces. It only contains one method `increment(Object record)` which defines how to increment the watermark for a given record.

Another class called `WatermarkInterval` contains the logic for maintaining low and high watermarks. The corresponding changes to `WorkUnit` and `WorkUnitState` have been made. Since this pull request mainly focuses on defining the interfaces, no migration code has been done to the framework yet.

zliu41 commented on gobblin-api/src/main/java/gobblin/source/extractor/Watermark.java:
```diff
@@ -0,0 +1,11 @@
+import org.codehaus.jackson.annotate.JsonTypeInfo;
+
+@JsonTypeInfo(use=JsonTypeInfo.Id.CLASS, include=JsonTypeInfo.As.PROPERTY, property="@class")
+public interface Watermark extends Comparable<Watermark>, Copyable<Watermark> {
```
There's already a Watermark interface in `gobblin.source.extractor.watermark`. Better to name this one `ComplexWatermark` or something like that.

liyinan926 replied:
This interface is better to be generic with a type parameter.

sahilTakiar replied:
The `increment` method is gone now, so I'm not sure if this still makes sense. A `Watermark` is no longer tied to a specific record type.

zliu41 commented on gobblin-api/src/main/java/gobblin/source/workunit/WorkUnit.java:
```diff
@@ -128,5 +150,21 @@ public void write(DataOutput out)
+    this.watermarkInterval.write(out);
+  }
+
+  public static class Factory {
```
Why not just make the constructor public?

zliu41 commented on gobblin-api/src/main/java/gobblin/configuration/SourceState.java:
```diff
@@ -67,6 +69,14 @@ public SourceState(State properties, List<WorkUnitState> previousTaskStates) {
     this.previousTaskStates.addAll(previousTaskStates);
   }
 
+  public List<WatermarkInterval> getPreviousWatermarkIntervals() {
```
At least for the kafka adapter, this method is not useful, since it only returns a list of previous watermarks from previousTaskStates, but we don't know which watermark interval corresponds to which kafka partition.

It will be better to return `Map<WorkUnitState, WatermarkInterval>`.

sahilTakiar replied:
Yeah, that makes sense. I actually just defined to remove this method altogether because there is already a `SourceState.getPreviousWorkUnitStates()` method, and a `WorkUnitState.getWatermarkInterval`, so this method isn't really necessary.

zliu41 commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  public boolean isTargetReached() {
+    return this.actualHighWatermark.equals(this.expectedHighWatermark);
```
Should an exception be thrown if `actualHighWatermark > expectedHighWatermark`?

liyinan926 replied:
Or an `IllegalStateException` should be thrown in `increment` if the high watermark is already going beyond the expected high watermark?

sahilTakiar replied:
Hmm, I'm not sure if we should make this an IllegalState. While I can't think of a concrete example, there may be users who want to pull data past the expected high watermark. For example, if the expected high watermark can't really be known until runtime, or if the `WorkUnit` pulls data until the maximum offset is reached.

chavdar commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java:
```diff
@@ -0,0 +1,91 @@
+      return this;
+    }
+
+    public WatermarkInterval build() throws CopyNotSupportedException {
```
Why do we need the exception?

liyinan926 commented on gobblin-api/src/main/java/gobblin/configuration/WorkUnitState.java:
```diff
@@ -171,14 +186,16 @@ public State getPreviousTableState() {
       throws IOException {
-    workunit.readFields(in);
+    this.workunit.readFields(in);
+    this.watermarkInterval.readFields(in);
```
This means another state store migration.

sahilTakiar replied:
Yeah. Since we probably want to avoid another state-store migration, I added a hack to ensure the serialization still works, but no state-store migration is necessary.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java:
```diff
@@ -0,0 +1,91 @@
+
+import com.google.common.base.Preconditions;
+
+public class WatermarkInterval implements Writable {
```
Like `Watermark`, I think this should be generic with a type parameter.

chavdar commented on gobblin-api/src/main/java/gobblin/source/extractor/Watermark.java:
```diff
@@ -0,0 +1,11 @@
+@JsonTypeInfo(use=JsonTypeInfo.Id.CLASS, include=JsonTypeInfo.As.PROPERTY, property="@class")
+public interface Watermark extends Comparable<Watermark>, Copyable<Watermark> {
+
+  public void increment(Object record);
```
I don't think I understand the point of this method. Which component is supposed to use it? This does not look like something many components should care about. The components which need to manipulate the checkpoint also should know what the implementation is. In that case, putting it in the interface is meaningless. 

Can you please make sure you add javadocs to public interfaces and methods?

chavdar commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java:
```diff
@@ -0,0 +1,91 @@
+    }
+  }
+
+  public void increment(Object record) {
```
See above comment for Watermark.increment()

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java:
```diff
@@ -0,0 +1,91 @@
+  private WatermarkInterval(Watermark lowWatermark, Watermark expectedHighWatermark) throws CopyNotSupportedException {
+    this.lowWatermark = lowWatermark;
+    this.expectedHighWatermark = expectedHighWatermark;
+    this.actualHighWatermark = lowWatermark.copy();
```
I suggest having a protected abstract method for initializing the `actualHighWatermark` instead of requiring an `Watermark` to be a `Copyable`. Then implementation classes can provide the right behavior for initializing it.

sahilTakiar replied:
Do you mean adding an protected abstract method to `Watermark` or `WatermarkInterval`? I wanted to keep `WatermarkInterval` as a concrete class.

sahilTakiar commented:
As per conversation with Chavdar. I have removed the `increment(Object record)` from the `Watermark` interface. It is now up to the `Extractor` class to update the `Watermark`. I have also removed the Jackson based serialization of the `Watermark` class and have just added `toJson` and `initFromJson` methods to the `Watermark` interface. Jackson is not that efficient, and can be messy to deal with, so its best use a simpler serialization approach.

So now the expected e2e flow is as follows:

1: The `Source` class creates a series of `WorkUnit`s, each `WorkUnit` contains a `WatermarkInterval`. The `WatermarkInterval` contains a low watermark, expected high watermark, and an actual high watermark. It is possible that a `WorkUnit` is not always able to pull data from its low watermark, to its high watermark. Take Camus for example, where map tasks only pull data for a (topic, partition) for "x" amount of time, and then move on to the next (topic, partition).

2: Each `WorkUnit` is then serialized into a SequenceFile

3: Each map task reads a series of `WorkUnit`s from this sequence file, and de-serializes them. It then executes these `WorkUnit`s.

4: The `Extractor` class is responsible for maintaining the following invariant. After each call to `readRecord()` the `Watermark` returned by `getCurrentHighWatermark` should cover all records up to and including the last record returned by `readRecord()`. This method may be called multiple times in order to report the % completion of the `WorkUnit`.

The following questions are still open:

1: I have added a new method to `Extractor` which is causing a lot of other classes to break

2: In order to avoid another state-store migration, I have added a hack to serialize the WatermarkInterval class. Should be discuss migrating away from SequenceFile formats?

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/Watermark.java:
```diff
@@ -0,0 +1,29 @@
+ */
+public interface Watermark extends Comparable<Watermark>, Copyable<Watermark> {
+
+  public void initFromJson(String json);
```
`fromJson` is sufficient.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/Watermark.java:
```diff
@@ -0,0 +1,29 @@
+   * {@link Watermark} values, where the lowWatermark is the starting point, and the highWatermark is the goal, what
+   * is the percent completion of this {@link Watermark}.
+   */
+  public float calculatePercentCompletion(Watermark lowWatermark, Watermark highWatermark);
```
Instead of returning a `float` between `[0.0, 1.0]`. Is it better to return a `short` between `[0, 100]`?

liyinan926 commented:
Wants to share some of my thoughts on the `Watermark` and `WatermarkInterval` classes:
1. `Watermark` should be generic with a type parameter and a method `public T getValue()`. 
2. `WatermarkInterval` should be immutable and only contain two things, the low and high `Watermark`s. Then it's essentially a `PairWritable`.
3. The logic of maintaining a `currentWatermark` and manipulating it should be left to the extractors. The runtime is only responsible for serialization/deserialization of the interval.
4. `Watermark` does not need to extend `Copyable` given the above three points.

zliu41 commented:
If `Watermark` is generic, then in addition to a `MyWatermark` class, I also need another class `MyWatermarkValueType` and do
`public class MyWatermark implements Watermark<MyWatermarkValueType>`, which doesn't seem necessary. Why not put those values directly in `MyWatermark`?

It seems the `getX` method can be implemented in each individual Watermark class if needed, e.g., `IntWatermark.getInt()`, `LongWatermark.getLong()`, `PairLongWatermark.getFirst()`, `PairLongWatermark.getSecond()` etc.

liyinan926 commented:
I don't quite understand your point. Regardless of whether `Watermark` is generic or not, you need to put some type of value into your `MyWatermark` class. With generic, the only additional thing you need to do is to specify explicitly what is the type of that value.

liyinan926 commented:
It's good to have a common method like `getValue` so it is able to get the value of a `Watermark` as long as it has access to the interface. If every implementation classes implement their own way of getting the value of a `Watermark`, then the code that needs the value needs to know the specific implementation class and very likely need to do a casting before it is able to call that method.

sahilTakiar commented:
@liyinan926 for points 2 and 3. I like the idea of having a `WatermarkInterval` only containing two `Watermark`s, a low and a high `Watermark`. However, I think it is still important that the runtime high watermark still get serialized along with the `WorkUnitState`. The source class should have access to the `WorkUnitState`s of the previous run, along with some way to access the low, expected high, and actual high watermark.

Perhaps, the `WorkUnit` can have a `WatermarkInterval` that contains the low and expected high `Watermark`, and the `WorkUnitState` can have the actual high `Watermark`. Thoughts?

liyinan926 commented:
@sahilTakiar, yeah, I think that's a good solution. So basically the source/extractor gives the `WorkUnitState` a reference to the `actualHighWatermark`, and increments that as records get extracted. Upon completion of the task, the `actualHighWatermark` and the `WatermarkInterval` get serialized as part of the `WorkunitState`. When serializing the `actualHighWatermark`, the `WorkUnitState` simply calls the `toJson` method.

sahilTakiar commented:
Updated. The `Watermark` interface now contains a `fromJson(JsonElement json)` and a `JsonElement toJson()` method.

I removed the new method from the `Extractor` interface so we don't break all the dependent classes, the `WorkUnitState.setHighWatermark(Watermark)` method is now responsible for setting the runtime high watermark. This method must be called during the `Extractor` initialization so that the `WorkUnitState` can properly report the progress of the `Watermark`.

I also changed `WatermarkInterval` to contain just two `Watermarks`, and now the `WorkUnit` serializes a `WatermarkInterval`, and the `WorkUnitState` just serializes a single high `Watermark`.

sahilTakiar commented:
@liyinan926 I am not entirely sure I understand why / how to make the `Watermark` generic. The `Watermark` interface may contain a variety of data structures used to track the progress of a `WorkUnitState` - e.g. maps, arrays, lists, etc. The only requirement is that all these data structures can be converted into a `JsonElement`, and that it is possible to do comparisons.

liyinan926 commented:
@sahilTakiar. What I thought is `Watermark` is a container for some value of a certain type, which can be parameterized. I also thought `Watermark` would support getting the internal value it contains. I think an interface for classes that may contain any types of values deserve being type parameterized. I am thinking of the following interface. 

```
@param <T> type of the watermark value
public interface Watermark<T> {
  
  public T getValue();
}
```

With this interface, you can have something like:

```
public class NumericalWatermark<T extends Numeric> implements Watermark<T> {
}
```

Like what I said above, adding a type parameter allows the compiler to do type checking and helps reduce possible abuse of `Object` like the following:

```
public class SimpleWatermark implements Watermark {
  public SimpleWatermark(Object obj) {
    ...
  }
}
```

sahilTakiar commented:
Hey, so I think the `Watermark` interface could be used in a couple different ways, depending on the source:

* For MySQL, each table has a single `Date` value inside its `Watermark` interface
* For SFTP, the `Watermark` interface contains a list of files that have already been pulled
* For Kafka, the `Watermark` contains a `Map` of from a Kafka (topic, partition) pair to some offset
 
For all the above use cases I think have a `getValue()` function would be good, for MySQL you would have `public class MySQL implements Watermark<Long>` and the method `public Long getValue()`.

However, if we take that approach then we add a restriction any class that implements the `Watermark` interface. The restriction is that the `Watermark` but be represented entirely by one data structure. While I can't think of any Source that would need multiple data structures to represent its `Watermark`, but it may entirely be possible.

Also, I think the `toJson()` method is already doing what the `getValue()` method would do. The `toJson()` method allows you to return the value of the `Watermark` in JSON format.

chavdar commented:
Let's discuss this in person. I don't understand why we need getValue(). The Watermark should be the watermark implementation. We already have a container for watermarks -- the workunit.

sahilTakiar commented:
@chavdar I still have some questions, so maybe we can sync up. Primarily, I realized that I will also have to change the `calculatePercentCompletion` method to take two `JsonElement`s instead of `Watermark`s, since this method will get called in the in the map task.

Which seems a little odd from a usability perspective, ideally the user should not need to worry about the fact that `Watermark` is getting serialized.

sahilTakiar commented:
Updated based on discussion with Chavdar.

@liyinan926 any more comments?

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java:
```diff
@@ -0,0 +1,66 @@
+ */
+public class WatermarkInterval {
+
+  private Watermark lowWatermark;
```
Both can be final.

sahilTakiar replied:
Fixed.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java:
```diff
@@ -0,0 +1,66 @@
+  private Watermark lowWatermark;
+  private Watermark expectedHighWatermark;
+
+  public static final String LOW_WATERMARK_TO_JSON_KEY = "low.watermark.to.json";
```
Constants better to be put at the top.

sahilTakiar replied:
Fixed.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java:
```diff
@@ -0,0 +1,66 @@
+    this.expectedHighWatermark = expectedHighWatermark;
+  }
+
+  public static class Builder {
```
Do you really need a builder class for this?

sahilTakiar replied:
No, not really, I removed it.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkSerializerHelper.java:
```diff
@@ -0,0 +1,37 @@
+ * way to serialize and de-serialize {@link Watermark}s, and is useful for implementing the {@link Watermark#toJson()}
+ * method.
+ */
+public class WatermarkSerializerHelper {
```
Seems this class is not used anywhere?

sahilTakiar replied:
It's not used anywhere. It just provides a default serialization technique for and users implementing a new Extractor or Source. It may become useful when using the complex watermarks in the Kafka Adapter.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/extractor/WatermarkSerializerHelper.java:
```diff
@@ -0,0 +1,37 @@
+   *
+   * @param jsonElement is a {@link JsonElement} that will be converted into a {@link Watermark}.
+   * @param clazz is the {@link Class} that the {@link JsonElement} will be converted into.
+   * @return an instance of a class that extends {@link Watermark}.
```
Should be `implements` instead of `extends`.

sahilTakiar replied:
Looks like in Java for generics, you use extends for interfaces also: http://stackoverflow.com/questions/18783916/check-if-a-generic-t-implements-an-interface

Replacing extends with implements throws a compile error

liyinan926 commented on gobblin-api/src/main/java/gobblin/configuration/WorkUnitState.java:
```diff
@@ -98,10 +105,56 @@ public void setWorkingState(WorkingState state) {
+     * internally in via a configuration key. Once a state-store migration can be done, the {@link Watermark} can be
+     * stored as Binary JSON.
+     */
+    setProp(ConfigurationKeys.WORK_UNIT_STATE_COMPLEX_ACTUAL_HIGH_WATER_MARK_KEY, this.actualHighWatermark.toJson());
```
The key name is really long. Can it be shortened without `COMPLEX`?

sahilTakiar replied:
Updated

sahilTakiar commented:
Comments have been address @liyinan926

liyinan926 commented:
LGTM.

sahilTakiar commented:
@chavdar any other comments?

sahilTakiar commented:
Merged
